### PR TITLE
Incorporated queing of RecyclePlots into  RevealPlots process

### PR DESCRIPTION
**Pull request by fanganga**
Work on the **Nymeria of Ny Sar**/ **"The Rains of Castamere"** interaction revealed inconsistencies in the use of recyclePlots - some actions that could empty the plot deck, such as revealing plots in the plot phase and **Nymeria of Ny Sar**, queued a recyclePlots step while others such as **"The Rains of Castamere"** did not.

This PR:
 - causes the code that creates a WhenRevealed window to be queued to queue a recyclePlots step immediately after
 - removes recyclePlots steps that are now redundant
 - introduces integration tests for motivating cases and edge cases where I anticipated an incorrect approach to this issue causing problems (mainly **At Prince Doran's Behest** interactions)

### CODE REVIEW

File: server/game/gamesteps/eventwindow.js [hunk 1]
```diff
@@ -70,6 +70,12 @@ class EventWindow extends BaseStep {
 
         if (this.event.getConcurrentEvents().some((event) => event.name === 'onPlotRevealed')) {
             this.openAbilityWindow('whenrevealed');
+            this.event
+                .getConcurrentEvents()
+                .filter((event) => event.name === 'onPlotRevealed')
+                .forEach((event) =>
+                    this.game.queueSimpleStep(() => event.plot.controller.recyclePlots())
+                );
```
Comment (Patane97) [inline server/game/gamesteps/eventwindow.js:78]: This is almost correct; I would not recommend attaching this logic to the `EventWindow` class; instead this should be part of the `RevealPlots` class.

I believe the quickest/easiest way would be to, for each plot being revealed, create a "onPlotsRecycled" event, with the handler of that event simply calling `event.plot.controller.recyclePlots()`, and then `addChildEvent` that event to the "onPlotRevealed" event right after it was created.

File: server/game/gamesteps/revealplots.js [hunk 2]
```diff
@@ -61,7 +61,8 @@ class RevealPlots extends BaseStep {
         });
 
         for (let plot of plots) {
-            event.addChildEvent(new Event('onPlotRevealed', { plot: plot }));
+            let plotRevealedEvent = new Event('onPlotRevealed', { plot: plot });
+            event.addChildEvent(plotRevealedEvent);
```
Comment (Patane97) [inline server/game/gamesteps/revealplots.js:65]: I mention in the comment above on recycling to happen within RevealPlots; this is probably the spot for it (and you creating that `plotRevealedEvent` is almost there

**Review by Patane97 — CHANGES_REQUESTED** [review]
Definitely on the correct track with this, however the wrong place to implement that change. Comments should outline, and give us a discord message if you need any help! :)

Also absolutely loving the tests, nice work!!

File: server/game/gamesteps/eventwindow.js (hunk 1 above)
Comment (fanganga) [reply server/game/gamesteps/eventwindow.js:78]: I agree that making the change in the `RevealPlots` class rather than the `EventWindow` class is cleaner. Adding an event with a handler that invokes the `recyclePlots()` method as a child of the "onPlotRevealed" event when the "onPlotRevealed" event is created introduces sequencing issues with plots like **At Prince Doran's Behest**, triggering the recycle before the **When Revealed**. I'll have another look for a solution that gets around the sequencing issues without having to put plot-specific code in  the `EventWindow` class, but probably not today.

Comment (Patane97) [reply server/game/gamesteps/eventwindow.js:78]: Yeah that makes sense; probably the wrong place then, but if you `thenExecute` it at the end of the '_TOP_LEVEL_REVEAL_' event (notably, after the `openInterruptWindowForAttachedEvents` is called), that should be equivalent of doing it at the same step that's happening now, but will also cover scenarios where `RevealPlots` is called outside of the plot phase.

Comment (Patane97) [reply server/game/gamesteps/eventwindow.js:78]: @fanganga I think your above suggestion is good enough to go with, especially as it also fixes that additional interaction you mention with cards like Bitterbridge + Cavalry.

Only thing I'd confirm is whether this will queue multiple plot shuffled for 1 player if they revealed multiple times (eg. will Nysar > Behest > *something* result in 3 recyclePlot triggers?). If so, just fix that and I'm happy to approve 💪

Comment (fanganga) [reply server/game/gamesteps/eventwindow.js:78]: Just tested, and it looks good
PlotRecycle

**Comment by Patane97** [comment]
@fanganga is there anything remaining for this one, or does it require a re-review & approval?

**Review by Patane97 — APPROVED** [review]
